### PR TITLE
Feature: digitalocean_image

### LINKED ISSUE #160: digitalocean_image resource?
This provider has a `digitalocean_image` data source, but since recently, it's also
supported to create custom images on digitalocean (https://developers.digitalocean.com/documentation/v2/#create-a-custom-image)

Would it be possible to expose this api as a `resource digitalocean_image`, similar to

https://www.terraform.io/docs/providers/google/r/compute_image.html
https://www.terraform.io/docs/providers/aws/r/ami.html

?

### PR DESCRIPTION
**Pull request by frezbo**
Signed-off-by: Noel Georgi <git@frezbo.com>

Fixes #160

Good to Have:

- The `Update` operation currently only updates the name since the `godo` client only supports that at the moment
- Copying to multiple regions is not supported yet

### CODE REVIEW

**Comment by frezbo** [comment]
Output of acceptance test:

```bash
❯ TF_ACC=1 DIGITALOCEAN_TOKEN="<redacted>" go test github.com/digitalocean/terraform-provider-digitalocean/digitalocean -run ^TestAccDigitalOceanImageFull$
ok      github.com/digitalocean/terraform-provider-digitalocean/digitalocean    (cached)
```

File: digitalocean/resource_digitalocean_image.go [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+// Ref: https://developers.digitalocean.com/documentation/v2/#retrieve-an-existing-image-by-id
+const (
+	imageAvailableStatus = "available"
+	imageDeletedStatus   = "deleted"
+)
+
+func resourceDigitalOceanImage() *schema.Resource {
```
Comment (andrewsomething) [inline digitalocean/resource_digitalocean_image.go:23]:
```suggestion
func resourceDigitalOceanCustomImage() *schema.Resource {
```

The methods names as well.

**Review by andrewsomething — CHANGES_REQUESTED** [review]
@frezbo, this looks great! I've left a few small comments in-line. 

Also, it would be great to get some docs for this. You can look at one of the existing pages for an exmaple: https://github.com/digitalocean/terraform-provider-digitalocean/blob/master/docs/resources/droplet_snapshot.md

File: digitalocean/provider.go [hunk 2]
```diff
@@ -110,6 +110,7 @@ func Provider() *schema.Provider {
 			"digitalocean_volume_attachment":                     resourceDigitalOceanVolumeAttachment(),
 			"digitalocean_volume_snapshot":                       resourceDigitalOceanVolumeSnapshot(),
 			"digitalocean_vpc":                                   resourceDigitalOceanVPC(),
+			"digitalocean_image":                                 resourceDigitalOceanImage(),
```
[the remark this answers is not in the document]
Comment (frezbo) [reply digitalocean/provider.go:113]: sure, I could make this change. `digitalocean_custom_image` makes more sense

File: digitalocean/resource_digitalocean_image_test.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	"strconv"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/acctest"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+)
+
+func TestAccDigitalOceanImageFull(t *testing.T) {
+	rString := acctest.RandString(5)
```
[the remark this answers is not in the document]
Comment (frezbo) [reply digitalocean/resource_digitalocean_image_test.go]: resolved

File: digitalocean/provider.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (frezbo) [reply digitalocean/provider.go]: resolved

File: digitalocean/resource_digitalocean_image.go [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+			"description": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.NoZeroValues,
+			},
+			"distribution": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "Unknown",
+				ValidateFunc: validation.NoZeroValues,
```
[the remark this answers is not in the document]
Comment (frezbo) [reply digitalocean/resource_digitalocean_image.go]: resolved

File: digitalocean/resource_digitalocean_image.go (hunk 1 above)
Comment (frezbo) [reply digitalocean/resource_digitalocean_image.go]: resolved

**Comment by frezbo** [comment]
Output of acceptance test after the changes:

```bash
❯ TF_ACC=1 DIGITALOCEAN_TOKEN="<redacted>" go test github.com/digitalocean/terraform-provider-digitalocean/digitalocean -run ^TestAccDigitalOceanCustomImageFull$
ok      github.com/digitalocean/terraform-provider-digitalocean/digitalocean    290.440s
```

**Review by andrewsomething — APPROVED** [review]
:+1:  Looks great. Thanks for the contribution!

**Comment by paulo-digital** [comment]
Nice feature introduction, it was missing. Although some important properties are missing here, the most notable are `size` and `user_data`, for a custom image those are a SHOULD HAVE property. Is unusable for my use cases, I hope to see those implemented in the future. Thank you, nice job.

**Comment by andrewsomething** [comment]
Hi @paulo-digital,

This resource just creates a custom image that can be used to create Droplets. You would need to specify the `size` and `user_data` when you create the Droplet, e.g:

```
resource "digitalocean_custom_image" "focal" {
  name         = "ubuntu-20.04-minimal-cloudimg-amd64"
  distribution = "Ubuntu"
  description  = "testing..."
  url          = "https://cloud-images.ubuntu.com/minimal/releases/focal/release/ubuntu-20.04-minimal-cloudimg-amd64.img"
  regions      = ["nyc3"]
}

resource "digitalocean_droplet" "example" {
  image     = digitalocean_custom_image.focal.id
  name      = "example-1"
  region    = "nyc3"
  size      = "s-1vcpu-1gb"
  user_data = "my script here..."
  ssh_keys  = [123456]
}
```

**Comment by paulo-digital** [comment]
Nice feature introduction, it was missing. But how to proper implement it? I tried something like below with no success. Can you lead me to a more complete documentation?   Thank you.

Declare:
```
resource "digitalocean_custom_image" "flatcar" {
  name   = "flatcar"
  url = "https://stable.release.flatcar-linux.net/amd64-usr/2605.7.0/flatcar_production_digitalocean_image.bin.bz2"
  regions = ["nyc3"]
}
```

Use?!:

```
resource "digitalocean_droplet" "quickstart_node" {
  name               = "${var.prefix}-quickstart-node"
  image              = digitalocean_custom_image.flatcar.name   
  region             = var.do_region
  size               = var.droplet_size`
}
```

**Comment by paulo-digital** [comment]
> Hi @paulo-digital,
> 
> This resource just creates a custom image that can be used to create Droplets. You would need to specify the `size` and `user_data` when you create the Droplet, e.g:
> 
> ```
> resource "digitalocean_custom_image" "focal" {
>   name         = "ubuntu-20.04-minimal-cloudimg-amd64"
>   distribution = "Ubuntu"
>   description  = "testing..."
>   url          = "https://cloud-images.ubuntu.com/minimal/releases/focal/release/ubuntu-20.04-minimal-cloudimg-amd64.img"
>   regions      = ["nyc3"]
> }
> 
> resource "digitalocean_droplet" "example" {
>   image     = digitalocean_custom_image.focal.id
>   name      = "example-1"
>   region    = "nyc3"
>   size      = "s-1vcpu-1gb"
>   user_data = "my script here..."
>   ssh_keys  = [123456]
> }
> ```

Fantastic, that's what I was trying to figure out!! Thank you.

**Pull request merged**